app/models/tuition.rb

```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class Tuition < ApplicationRecord
+  # search
+  include DynamicSearch
+
+  # dynamic_search
+  search_keys :academic_term_id
+
+  # relations
+  belongs_to :academic_term
+  has_many :unit_tuitions, dependent: :destroy
+  has_many :units, through: :unit_tuitions
+
+  # validations
+  validates :fee, numericality: { greater_than_or_equal_to: 0 }
+  validates :foreign_student_fee, numericality: { greater_than_or_equal_to: 0 }
+  validate :fee_must_be_specified
+  validates_associated :unit_tuitions
+
+  def fee_must_be_specified
+    errors.add(:base, :not_specified_fees) if [fee, foreign_student_fee].map(&:zero?).all?
```

Comment:
Burada, her iki alanı da zorunlu tutsak mantıklı olmaz mı? Hangi birimde yabancı uyruklu öğrenci var veya olacak bunu kestirmek zor, bu durum sistemde hata oluşmasına neden olur.

Ek olarak:

```ruby
[fee, foreign_student_fee].all?(&:zero?)
```

bu şekilde map'e gerek olmadan kullanılabilir.

Reply:
Sadece yabancı öğrenciler için ücretlendirme veya tam tersi olabilir diye bu şekilde yapmıştım. Her harç bilgisi için bu iki alan zorunlu doldurulması gerekiyor mu?

Reply:
Senin dediğin şekilde yapılacak olsaydı, `foreign_student_fee` gibi birşey olmamalıydı. Borçlandırma türü veya borçlandırma koşulu eklemek gerekirdi.

Bizim burada kontrol etmemiz gereken iki durum var;

1. Tüm programlar için harç tutarı girildi mi?
2. Yabancı uyruklu öğrencisi olan programlar için harç tutarı girildi mi? (Bunu tahmin etmek bence zor, o yüzden Tüm programlar için yabancı uyruklular için harç tutarı girildi mi? gibi bir kontrol yapmamız gerekir.) Bu yüzden harç tutarlarında iki alana da değer girmeli?

Girilmesi gereken verilerde hata olursa borç oluşturma ve ödeme tarafında sorun çıkar. Bu durumun bir yerde kontrol edilmesi gerekli. Borç oluşturma esnasında kontrol edebiliriz belki ama borç oluşturduktan sonra öğrenci kayıtlanması olabilir bu durumda otomatik borç ataması filan yapması gerekli o zamanlarda sorun çıkabilir.

Reply:
Birinci öğretim programlarında genel olarak harç borcu yok. Daha doğrusu var ancak devlet tarafından karşılanıyor. O nedenle öğrenciye yansıyan bir borç yok. Yabancı uyruklu öğrenciler için ise bakanlar kurulu kararına göre normal program öğrenim ücretinin 3-5 katı aralığında harç borcu çıkartılması gerekiyor. Biraz önce telefonla sorduğumda tüm birimlerde yabancı uyruklular için bir şekilde harç borcu ataması yapılıyormuş. Yani hatalı girişleri engellemek için yabancı uyruklulara  ">0" kontrolü yapılabilir.  Ancak "harç ücreti" alanı 0 olabilir. Bundan bağımsız olarak harç ücretlerinin girişinden sonra toplu bir raporlama şeklinde de yapılabilir (Harç ücreti atanmayan birimler vs)